From f8289e595a8dc1b8c9738833ffa3d5a98b711268 Mon Sep 17 00:00:00 2001
From: Remi Gacogne <remi.gacogne@powerdns.com>
Date: Mon, 31 Aug 2020 10:25:04 +0200
Subject: [PATCH] auth: Handle the extra single-row result set of MySQL stored
 procedures

---
 modules/gmysqlbackend/smysql.cc | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/modules/gmysqlbackend/smysql.cc b/modules/gmysqlbackend/smysql.cc
index 7408742..23c4841 100644
--- a/modules/gmysqlbackend/smysql.cc
+++ b/modules/gmysqlbackend/smysql.cc
@@ -303,7 +303,10 @@ public:
         d_resnum = mysql_stmt_num_rows(d_stmt);
         // XXX: For some reason mysql_stmt_result_metadata returns NULL here, so we cannot
         // ensure row field count matches first result set.
-        if (d_resnum > 0) { // ignore empty result set
+        // We need to check the field count as stored procedure return the final values of OUT and INOUT parameters
+        // as an extra single-row result set following any result sets produced by the procedure itself.
+        // mysql_stmt_field_count() will return 0 for those.
+        if (mysql_stmt_field_count(d_stmt) > 0 && d_resnum > 0) { // ignore empty result set
           if (d_res_bind != nullptr && (err = mysql_stmt_bind_result(d_stmt, d_res_bind))) {
             string error(mysql_stmt_error(d_stmt));
             releaseStatement();
